### PR TITLE
OCPBUGS-33275: extramanifest configmaps validation update

### PR DESCRIPTION
**Pull request by Missxiaoguo**
# Background / Context
LCA performs Kubernetes Dry-run to validate the manifests included in the extraManifests configmaps. Any invalid/webhook badRequest type of error returned from dry-run fails the Prep stage and blocks the upgrade.

# Issue / Requirement / Reason for change
Kubernetes Dry-run sends requests to API server, where the server processes the request as if it were real, but without acutally making any modifications to the actual resources. As all API requests are single-resource requests, dryrun doesn't work for resources that depends on eariler resources that are not persisted in the cluster. For example, in a multi-object manifest that includes namespace A and resource A created within the namespace A. Running a dry-run on resource returns the namespace not found error.

The extra manifests are configurations intented to be applied to the new cluster version. Runing dry-run on the old version aims to thoroughly check for possible manifest error before upgrading. However, in cases where extramanifests configmaps contain the dependent resources, the Dry-run failure results in failed Prep, but the failure may be not the case for the new cluster version.

# Solution / Feature Overview
Not only in the case of dependent resources, but there may also be other dry-run errors that occur only on the old cluster version.  The dry-run validation should be a soft validation. The errors such as invalid or webhook badRequest type of error from dry-run are considered warnings. This also includes cases where CRDs are missing from the old version and dependent namespace does not exist on the old version but is also not found in the configmaps. The warnings do not block upgrade. Instead, the Prep status condition could be updated with a warning message and the failure details can be provided in an IBU annotation. The other type of errors such as random chars, missing resource Kind, resource ApiVersion, resource name or the presence of disallowed resource types, are treated as critical errors which will fail the Prep.

# Implementation Details

# Other Information
- If user uses the ZTP/TALM to run IBU with extra manifests configmaps, the prep stage policy should include `message` check with "Prep stage completed successfully" which would prevent CGU to complete if there's validation warning

- Log error in the upgrade handler for the error that causes the upgrade to fail.

/cc @donpenney @pixelsoccupied @leo8a

### CODE REVIEW

File: internal/extramanifest/extramanifest.go [hunk 1]
```diff
@@ -179,79 +192,149 @@ func ApplyExtraManifest(ctx context.Context, dc dynamic.Interface, restMapper me
...
-func (h *EMHandler) ValidateExtraManifestConfigmaps(ctx context.Context, content []ibuv1.ConfigMapRef, ibu *ibuv1.ImageBasedUpgrade) error {
+// ValidateExtraManifestConfigmaps validates IBU extramanifest configmaps and returns validation warning and error.
+// It iterates over all manifests included in the configmaps and perform Kubernetes Dry-run.
+// Errors from Dry-run apply are considered warnings. This also includes cases where CRDs are missing from the current
+// stateroot and dependent namespace does not exist on the current stateroot but is also not found in the configmaps.
+// Other validation failures are treated as errors, such as random chars, missing resource Kind, resource ApiVersion,
+// resource name or the presence of disallowed resource types
+func (h *EMHandler) ValidateExtraManifestConfigmaps(ctx context.Context, content []ibuv1.ConfigMapRef) (string, error) {
+	var errs = []string{}     // validation errors
+	var warnings = []string{} // validation warnings
```
Comment (leo8a) [inline internal/extramanifest/extramanifest.go:203]:
```suggestion
	var errs []string     // validation errors
	var warnings []string // validation warnings
```
safer empty slice declaration (according to [styleguide](https://google.github.io/styleguide/go/decisions#nil-slices))

File: internal/extramanifest/extramanifest.go [hunk 2]
```diff
@@ -47,18 +47,26 @@ import (
...
 
+//+kubebuilder:rbac:groups="",resources=namespaces,verbs=get;watch;create;
+
 const (
-	ExtraManifestPath           = "/opt/extra-manifests"
-	PolicyManifestPath          = "/opt/policy-manifests"
-	WarnAnnotationUnknownCRDKey = "lca.openshift.io/warn-extramanifest-cm-unknown-crd"
+	ExtraManifestPath                        = "/opt/extra-manifests"
+	PolicyManifestPath                       = "/opt/policy-manifests"
+	ExtraManifestValidationWarningAnnotation = "lca.openshift.io/extramanifest-cm-validation-warning"
```
Comment (leo8a) [inline internal/extramanifest/extramanifest.go:55]:
```suggestion
	Path                        = "/opt/extra-manifests"
	PolicyManifestPath                       = "/opt/policy-manifests"
	ValidationWarningAnnotation = "lca.openshift.io/extramanifest-cm-validation-warning"
```
optional nit: exported constants should not start with package name (see [styleguide](https://google.github.io/styleguide/go/decisions#constant-names))

in this way we could call it like `extramanifest.ValidationWarningAnnotation` instead of `extramanifest.ExtraManifestValidationWarningAnnotation` (repeated)

totally get it if we don't want to go over this refact in this PR, just suggesting it

File: internal/extramanifest/extramanifest.go [hunk 3]
```diff
@@ -179,79 +192,149 @@ func ApplyExtraManifest(ctx context.Context, dc dynamic.Interface, restMapper me
...
+						h.Log.Info(fmt.Sprintf("Created namespace %s for dryrun", ns.Name))
+						fakedNamespacesForDryrun[ns.Name] = true
+					} else {
+						return validationWarn, fmt.Errorf("failed to query namespace %s: %w", ns, err)
+					}
+				}
+			}
+
+			// Apply manifest with dryRun
+			if err := ApplyExtraManifest(ctx, h.DynamicClient, h.Client.RESTMapper(), manifest, dryRun); err != nil {
```
Comment (leo8a) [inline internal/extramanifest/extramanifest.go:300]:
```suggestion
			if err := ApplyExtraManifest(ctx, h.DynamicClient, h.Client.RESTMapper(), manifest, true); err != nil {
```
`dryRun` is always true, maybe could be simplified (if so, there's a declaration above that could be removed)

**Review by leo8a — COMMENTED** [review]
nice work Angie, thanks

dropped some minor comments below, mainly related to code style guidelines

**Comment by jc-rh** [comment]
/cc @sudomakeinstall2 

> If user uses the ZTP/TALM to run IBU with extra manifests configmaps, the prep stage policy should include message check with "Prep stage completed successfully" which would prevent CGU to complete if there's validation warning

File: controllers/upgrade_handlers.go [hunk 4]
```diff
@@ -226,7 +229,7 @@ func (u *UpgHandler) PrePivot(ctx context.Context, ibu *ibuv1.ImageBasedUpgrade)
 	err = u.RebootClient.RebootToNewStateRoot("upgrade")
 	if err != nil {
 		//todo: abort handler? e.g delete desired stateroot
```
Comment (pixelsoccupied) [inline controllers/upgrade_handlers.go:231]: i think this todo can be removed right?

File: docs/image-based-upgrade.md [hunk 5]
```diff
@@ -291,9 +291,11 @@ oc patch imagebasedupgrade upgrade -n openshift-lifecycle-agent --type='json' -p
...
 
 - Perform the following validations:
-  - If the oadpContent is populated, validate that the specified configmap has been applied and is valid
-  - If the extraManifests is populated, validate that the specified configmap has been applied and is valid
-    - If a required CRD is missing from the current stateroot, a warning message will be included in the IBU CRs with annotation `lca.openshift.io/warn-extramanifest-cm-unknown-crd`.
+  - If the `oadpContent` is populated, validate that the specified configmap has been applied and is valid
+  - If the `extraManifests` is populated, validate that the specified configmap has been applied and is valid
+    - Validation errors from Dry-run apply are treated as warnings. This also includes cases where CRDs are missing from the current stateroot and dependent namespace does not exist on the current stateroot but is also not found in the configmaps.
+    LCA doesn't block the upgrade due to validation warnings. Instead, it updates the Prep status condition with a warning message and annotates IBU with the annotation `lca.openshift.io.extramanifest-cm-validation-warning`, providing detailed information about the failures
```
Comment (pixelsoccupied) [inline docs/image-based-upgrade.md:297]: What do you think about updating annotation to something like
```
extra-manifest.lca.openshift.io/warn-validation-error
```

Follows the recommended pattern `subsystem.lca.ocp.io/name`  and matches our other ones we have (autorollback and upcoming HC one too)

File: controllers/upgrade_handlers.go [hunk 6]
```diff
@@ -428,7 +435,7 @@ func (u *UpgHandler) PostPivot(ctx context.Context, ibu *ibuv1.ImageBasedUpgrade
 
 	if err := u.RebootClient.DisableInitMonitor(); err != nil {
 		// Don't fail the upgrade on failure here, just log it
-		u.Log.Error(err, "unable to disable LCA init monitor")
+		u.Log.Error(err, "Unable to disable LCA init monitor")
```
Comment (pixelsoccupied) [inline controllers/upgrade_handlers.go:438]: i think all errors are recommended to start with lowercase letter?

File: internal/extramanifest/extramanifest.go [hunk 7]
```diff
@@ -179,79 +192,149 @@ func ApplyExtraManifest(ctx context.Context, dc dynamic.Interface, restMapper me
...
-			case errors.As(err, &groupDiscoveryErr) || meta.IsNoMatchError(err): // Capture unknown CRD issue
-				key := manifest.GroupVersionKind().String()
-				if _, exists := unknownCRDs[key]; !exists {
-					unknownCRDs[key] = manifest.GetName()
-					h.Log.Info("unknown CRD", "CRD", key)
+			// Verify disallowed resource
+			if kinds, exists := unSupportedResources[mApiVersion]; exists {
+				if lo.Contains(kinds, mKind) {
+					errs = append(errs, fmt.Sprintf("%s[%s] in extramanifests is not allowed", mKind, mName))
+					continue
```
Comment (pixelsoccupied) [inline internal/extramanifest/extramanifest.go:263]: Ah nice..so we now collecting all the errors before returning yeah? (unlike prev we were returning early?)

File: internal/extramanifest/extramanifest.go [hunk 8]
```diff
@@ -179,79 +192,149 @@ func ApplyExtraManifest(ctx context.Context, dc dynamic.Interface, restMapper me
...
+	if len(warnings) == 0 && len(errs) == 0 {
+		h.Log.Info("ExtraManifests configmaps are validated", "configmaps", content)
+		return "", nil
 	}
 
-	h.Log.Info("ExtraManifests configmaps are validated", "configmaps", content)
-	return nil
+	if len(warnings) > 0 {
+		validationWarn = strings.Join(warnings, "; ")
+		h.Log.Info(fmt.Sprintf("[WARNING] ExtraManifests configmaps validation: %s", validationWarn))
```
Comment (pixelsoccupied) [inline internal/extramanifest/extramanifest.go:321]: I wish we had `.log.Warning` option!

File: internal/extramanifest/extramanifest.go [hunk 9]
```diff
@@ -179,79 +192,149 @@ func ApplyExtraManifest(ctx context.Context, dc dynamic.Interface, restMapper me
...
+			// Add to namespace set when manifest is Namespace
+			if mKind == "Namespace" {
+				nsManifestsSet[mName] = true
+				continue
+			}
+
+			// Check if manifest's namespace exists on cluster before running dryrun
+			if mNamespace != "" && mNamespace != "default" {
+				ns := &corev1.Namespace{}
+				if err := h.Client.Get(ctx, types.NamespacedName{Name: mNamespace}, ns); err != nil {
```
Comment (pixelsoccupied) [inline internal/extramanifest/extramanifest.go:276]: What do you think about moving the fakeNS creation + deletion moved out of ValidateExtraManifestConfigmaps? It was a bit hard to follow

File: internal/extramanifest/extramanifest.go [hunk 10]
```diff
@@ -179,79 +192,149 @@ func ApplyExtraManifest(ctx context.Context, dc dynamic.Interface, restMapper me
 	return nil
 }
 
-func (h *EMHandler) ValidateExtraManifestConfigmaps(ctx context.Context, content []ibuv1.ConfigMapRef, ibu *ibuv1.ImageBasedUpgrade) error {
+// ValidateExtraManifestConfigmaps validates IBU extramanifest configmaps and returns validation warning and error.
+// It iterates over all manifests included in the configmaps and perform Kubernetes Dry-run.
+// Errors from Dry-run apply are considered warnings. This also includes cases where CRDs are missing from the current
```
Comment (pixelsoccupied) [inline internal/extramanifest/extramanifest.go:197]: I think should say something like "Most errors from Dry-run apply..." ?

**Review by pixelsoccupied — COMMENTED** [review]
Just tiny comments! But lgtm otherwise

File: controllers/upgrade_handlers.go (hunk 6 above)
Comment (donpenney) [reply controllers/upgrade_handlers.go:438]: This would be an error log, not an error that can be wrapped

File: internal/extramanifest/extramanifest.go (hunk 7 above)
Comment (Missxiaoguo) [reply internal/extramanifest/extramanifest.go:263]: Exactly, collect all errors and warnings.

File: internal/extramanifest/extramanifest.go (hunk 10 above)
Comment (Missxiaoguo) [reply internal/extramanifest/extramanifest.go:197]: yeah, I think I should make it more accurate

File: internal/extramanifest/extramanifest.go (hunk 9 above)
Comment (Missxiaoguo) [reply internal/extramanifest/extramanifest.go:276]: Certainly, that's good suggestion to simplify this validation function

Comment (Missxiaoguo) [reply internal/extramanifest/extramanifest.go:292]: Unable to completely moving the creation part out but I hope this version is easier to follow
